salmon/model.py

```diff
@@ -129,20 +129,71 @@ def __init__(self, explanatory, response, intercept=True):
-                variable.
+            explanatory - Explanatory variables used to condition the model on.
+                Represented either by an `Expression` that is either a single 
+                term or a `Combination` of terms, or data in the form of a 
+                pandas `DataFrame` or numpy 1-D / 2-D `ndarray`. 
+            response - Response variables used as the target for modeling. 
+                Represented either by an `Expression` object, or data in the
+                form of a pandas `Series` or numpy 1-D `ndarray` object.
             intercept - A boolean indicating whether an intercept should be
                 included (True) or not (False).
         """
+        # may be overwritten if `explanatory` and `response` are given as 
+        # data structures rather than expressions
+        self.training_data = None  
+        
         if explanatory is None:
             explanatory = 0
 
         if isinstance(explanatory, (int, float)):
             explanatory = Constant(explanatory)
+        
+        if (isinstance(explanatory, (pd.DataFrame, np.ndarray)) or 
+                isinstance(response, (pd.Series, np.ndarray))):
+
+            assert(isinstance(explanatory, (pd.DataFrame, np.ndarray)) and 
+                isinstance(response, (pd.Series, np.ndarray)))
+
+            if isinstance(explanatory, np.ndarray):
+                if len(explanatory.shape) == 1:
+                    explanatory = explanatory.reshape((len(explanatory), 1))
+                else:
+                    assert(len(explanatory.shape) == 2)
+
+                expl_data = explanatory
+                expl_cols = ["Ex_{}".format(i+1) for i in range(expl_data.shape[1])]
+            else:  # isinstance(explanatory, pd.DataFrame)
+                expl_data = explanatory.values
+                expl_cols = list(explanatory.columns)
+
+            if isinstance(response, np.ndarray):
+                if len(response.shape) == 1:
+                    response = response.reshape((len(response), 1))
+                else:
+                    assert(response.shape[1] == 1)
+                    assert(len(response.shape) == 2)
+
+                resp_data = response
+                resp_cols = ["Response"]
+            else:  # isinstance(response, pd.Series)
+                resp_data = response.values.reshape((len(response), 1))
+                if response.name is None:
+                    resp_cols = ["Response"]
+                else:
+                    resp_cols = [response.name]
+            
+            self.X_train_ = LightDataFrame(expl_data, columns=expl_cols)
+            self.y_train_ = LightDataFrame(resp_data, columns=resp_cols)[:, 0]
+            self.training_data = LightDataFrame(
+                np.hstack([expl_data, resp_data]),
+                columns=expl_cols+resp_cols,
+            )
+            
+            # Bypass construction because we know explicitly that the terms will not collide
```

Comment:
Do we need to do this? Worried that this can add to the runtime. How long does this take?

Reply:
I did a quick test. To use hstack on a randomly generated 1000x1000 matrix (X) and a 1000x1 matrix (Y), it takes on average 0.0235 seconds (note this already accounts for the time spent generating the matrices). This seems like a pretty minimal impact; although I think it may depend on what kind of margins we have between our speed and the others.